src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+    let eom = memory.data(&caller).len();
+    let Ok(_) = memory.grow(&mut caller.as_context_mut(), 1) else {
+        return 0;
+    };
+
+    // I need these two to be the same lifetime it seems
+    let (mem_data, store_data) = memory.data_and_store_mut(&mut caller);
+
+    let rb_buffer = &store_data.buffer;
+    let length = rb_buffer.len;
+
+    let pos_loc = eom + length * core::mem::size_of::<hb_glyph_info_t>();
+    let end_loc = pos_loc + length * core::mem::size_of::<GlyphPosition>();
+
+    // This _should_ work ..
+    mem_data[eom..pos_loc].copy_from_slice(bytemuck::cast_slice(&rb_buffer.info));
+    mem_data[pos_loc..end_loc].copy_from_slice(bytemuck::cast_slice(&rb_buffer.pos));
+
+    let buffer_contents = CBufferContents {
+        length: length as u32,
+        info: eom as u32,
+        position: pos_loc as u32,
+    };
+    let Ok(()) = memory.write(
+        &mut caller.as_context_mut(),
+        cbuffer as usize,
+        bytemuck::bytes_of(&buffer_contents),
+    ) else {
+        return 0;
+    };
+
+    1
+}
+
+// fn buffer_set_contents(buffer: u32, cbuffer: &CBufferContents) -> bool;
+// Copy the buffer_contents structure back into the host shaping engine's buffer. This should typically be called at the end of shaping.
+fn buffer_set_contents(mut caller: Caller<'_, ShapingData>, _buffer: u32, cbuffer: u32) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut buffer = [0; core::mem::size_of::<CBufferContents>()];
+    memory.read(caller.as_context_mut(), cbuffer as usize, &mut buffer);
+    let Ok(buffer) = bytemuck::try_from_bytes::<CBufferContents>(&buffer) else {
+        return 0;
+    };
+
+    let (mem_data, store_data) = memory.data_and_store_mut(&mut caller);
+
+    let array_length = buffer.length as usize * core::mem::size_of::<hb_glyph_info_t>();
+
+    store_data.buffer.len = buffer.length as usize;
+
+    store_data.buffer.info.clear();
+    store_data
+        .buffer
+        .info
+        .extend_from_slice(bytemuck::cast_slice(
+            &mem_data[buffer.info as usize..buffer.info as usize + array_length],
+        ));
+
+    store_data.buffer.pos.clear();
+    store_data
+        .buffer
+        .pos
+        .extend_from_slice(bytemuck::cast_slice(
+            &mem_data[buffer.position as usize..buffer.position as usize + array_length],
+        ));
+
+    1
+}
+
+// fn debugprint(s: *const u8);
+// Produces a debugging message in the host shaper's log output; the variants debugprint1 ... debugprint4 suffix the message with a comma-separated list of the integer arguments.
+// rust varargs when
+fn debugprint(mut caller: Caller<'_, ShapingData>, s: u32) {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    // seems reasonable?
+    let bytes = &memory.data(&caller)[s as usize..];
```

Comment:
this also use checked indexing as to not crash anything

Reply:
I don't think that's needed here. as `s` is gived to us by the module and is (probably?) guarabteed to be a valid index.

(Also this worked fine for the Calculator wasm, and printed it and everything.)

Reply:
I'm thinking of the case of a font that wants to crash the shaper.